### DISCUSSION TITLE
Is there a way to set the user for traces?

### DISCUSSION THREAD

**Discussion by kujtimiihoxha** [aws-powertools/powertools-lambda-typescript#1224] [Q&A] [answered]
I am trying to figure out how to set the userId for the traces based on https://docs.aws.amazon.com/xray/latest/devguide/xray-sdk-nodejs-segment.html.

But it seems like ` tracer.getSegment()` does not have the `setUser` method.

**Comment by dreamorosi** [comment, marked as answer]
Hi @kujtimiihoxha, thanks for opening the discussion & for using Powertools.

The Tracer utility uses the `Segment` and `Subsegment` objects from `aws-xray-sdk-core` and doesn't apply any changes to their prototype, so the functionalities are the same as the ones offered by the SDK.

The `getSegment()` method from Tracer can return either a `Segment` or a `Subsegment`. In AWS Lambda this means either the `facade` segment (the one created by AWS Lambda, the service, which is immutable), or any subsegment (which can be the one created by Powertools - i.e. `# index.handler` - or any other subsegment created by you).

As mentioned above, in Lambda, the main segment (of type `Segment`) is immutable, meaning that you cannot add annotation or other metadata to it. This is explained in the last paragraph of [this docs page](https://docs.aws.amazon.com/xray/latest/devguide/xray-services-lambda.html):
> The other difference between instrumenting Lambda functions and web applications is that the segment that Lambda creates and sends to X-Ray can't be modified by your function code. You can create subsegments and record annotations and metadata on them, but you can't add annotations and metadata to the parent segment.

If we look at the code of the `aws-xray-sdk-core` we can see that the `setUser` method is a member of the `Segment` class, not the `Subsegment` - [see here](https://github.com/aws/aws-xray-sdk-node/blob/master/packages/core/lib/segments/segment.d.ts#L24). 

To evidence this, we can call `getSegment` and then explicitly cast the result as `Segment`:
![image](https://user-images.githubusercontent.com/7353869/211807880-eae881c5-e9b6-4246-a80d-cc351e21c3d7.png)

I would **recommend against doing that** because the segment in Lambda should be immutable as discussed above.

This is also compatible with what stated in [the doc section](https://docs.aws.amazon.com/xray/latest/devguide/xray-sdk-nodejs-segment.html#xray-sdk-nodejs-segment-userid) that you linked in your OP:
> Record user IDs on request segments to identify the user who sent the request. This operation isn’t compatible with AWS Lambda functions because segments in Lambda environments are immutable. The setUser call can be applied only to segments, not subsegments.

So unless this is supported by the SDK, Powertools won't be able to support it.

Hope this helps!

**Comment by kujtimiihoxha** [comment]
@dreamorosi thanks it seems like I did not fully understand how x-ray and lambda work together thanks for the response.

I will just use annotations instead.
